### PR TITLE
New features, Bugfix...

### PR DESCRIPTION
**Pull request by xmurobi**
1. Multiple reader instances and one writer instance allow, reader will check the new data added by writer dynamically.
2. Better command parser for `show` command.
3. Agg function `Gap` to find the big hole in data base on `show` command.
4. More types support in read/write layer.
5. New slait support. 
6. Bugs fix.

### CODE REVIEW

**Comment by umitanuki** [comment]
It's a lot of work, thanks! Is it possible to split this into individual features/fixes so we can review more easily?

**Comment by xmurobi** [comment]
Hi @umitanuki, It takes time, but recently I'm working on my quant system(multiple subsystem and lots of work left). So if anyone can help for this, I'll be very appreciate. If no one can help, I'll do it when I have time. Also, I'll continue to modify this great project to meet my requirements and contribute code.

**Comment by umitanuki** [comment]
We can definitely help there, but I need your input of each topic. For example, "bug fix" of what issue?

**Comment by xmurobi** [comment]
So what should I do? Separate the push request to multiple ones and tag them more details one by one?

**Comment by umitanuki** [comment]
That would be awesome. You can start from just one, for bug fix if it's small enough.

File: Dockerfile [hunk 1]
```diff
@@ -12,6 +12,8 @@ FROM golang:alpine AS builder
 ARG tag=latest
 ENV DOCKER_TAG=$tag
 ENV GO111MODULE=on
+# Fuck GFW, need to proxy all traffic to make it done
+ENV https_proxy=http://192.168.0.101:61118
```
Comment (Notargets) [inline Dockerfile]: need this to dissappear

File: executor/readvariable.go [hunk 2]
```diff
@@ -1,21 +1,25 @@
...
-#cgo CFLAGS: -O3 -Wno-ignored-optimization-argument
+#cgo CFLAGS: -O3 -Wno-ignored-optimization-argument -std=c99
 */
 import "C"
 
 func (r *reader) readSecondStage(bufMeta []bufferMeta, limitCount int32, direction DirectionEnum) (rb []byte, err error) {
+	// fmt.Printf(">>>>DEBUG::meta %v, limitCount %d, direction %v\n", bufMeta, limitCount, direction)
+
+	result := []byte{}
 	/*
```
Comment (Notargets) [inline executor/readvariable.go]: what's going on here? "result" isn't used, but we build it?

File: executor/readvariable.go [hunk 3]
```diff
@@ -115,7 +121,8 @@ func (r *reader) readSecondStage(bufMeta []bufferMeta, limitCount int32, directi
 				}
 			}
 		}
-		rb = rb[:rbCursor]
+		// fmt.Printf("cursor %d, rb size %d, cut size %d\n", rbCursor, len(rb), len(rb[:rbCursor]))
+		result = append(result, rb[:rbCursor]...)
 		fp.Close()
```
Comment (Notargets) [inline executor/readvariable.go]: same as above - why build "result"?
BTW - this is in a critical performance routine - previously we found appending here is far too slow

File: executor/wal.go [hunk 4]
```diff
@@ -772,6 +779,10 @@ func (wf *WALFileType) cleanupOldWALFiles(rootDir string) {
 						os.Remove(filePath)
 					} else {
 						w, err := NewWALFile(rootDir, filePath)
+						if w == nil {
+							log.Info("WALFILE: %s is probably opened by other instances, jump over it...", filename)
+							continue
+						}
 						if err != nil {
```
Comment (Notargets) [inline executor/wal.go]: I think you'd reliably corrupt data this way - multiple writers are free to write over the same files...

**Comment by Notargets** [comment]
I just did a "cherry picking" merge of this - thank you for your contribution!

**Pull request closed without merging**